[systemdreceiver] Add support for scraping unit status

SquidDev opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This is start at implementing #33532. I'm trying to do this in relatively minor increments to make this reviewable, so this *only* adds support for reporting the active units' status.

This follows a similar approach to the httpcheck receiver, with an attribute for every possible state, with one metric set to `1`, and the rest set to `0`.

```
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="active"} = 1
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="reloading"} = 0
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="inactive"} = 0
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="failed"} = 0
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="activating"} = 0
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="deactivating"} = 0
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="maintenance"} = 0
systemd.unit.state{systemd.unit.name="nginx", systemd.unit.active_state="refreshing"} = 0
```

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Part of #33532, but possibly not enough to mark as fixed. 

#### Testing
I have done some manual testing, and confirmed this generates the metrics I'd expected. Aside from the auto-generated tests, I'm afraid I've not written any unit/integration tests — I wanted some advice here first, about whether it was better to try to mock out the dbus interface, or instead try to run the tests against a real systemd instance.

#### Documentation
I've updated the README of the systemdreceiver component to mention the metric exposed, and the configuration options.

SquidDev commented:
Have rebased on top of master, and tried to stub out the dbus connection, so we can write some tests for this. Coverage is now at the requested 80% :).

I'm aware this PR is now well over the requested 500 lines. The majority of this is generated code (or golden test data), so hopefully shouldn't be too bad to review, but let me know there's a way to split this up further.

nichenke commented:
Following along - definitely interested. I'm poking at the cgroups side of things locally -- not sure if we need to vendor in cadvisor like [awscontainerinsightreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a55f7f0b40699ad55a5b568f1f1d810b98f6ba7c/receiver/awscontainerinsightreceiver) or the simpler podman/k8s/docker stats methods will work.

atoulme commented on receiver/systemdreceiver/testdata/expected_metrics/metrics_golden.yaml:
empty file?

SquidDev commented:
For cgroups, I'm currently just using https://github.com/containerd/cgroups/ directy. Though this does make supporting cgroups v1 and v2 a bit more awkward — I'm currently just focusing on the latter. 

I'd actually been drawing more inspiration for the host receiver (as much as I can, sadly cgroups has a different model for a lot of things). So CPU usage mode (kernel vs user) is an attribute rather than having separate `container.cpu.usage.{kernel,user}mode`.

I'm not 100% sure this is the right approach, so thoughts welcome!

SquidDev committed:
[systemdreceiver] Scraping unit status

SquidDev committed:
[systemdreceiver] Fix go.mod

SquidDev committed:
[systemdreceiver] Expose unit name as a resource attribute

 - Store possible unit states in a slice, to ensure metrics are emitted
   in a deterministic order.

SquidDev commented:
Have rebased on top of master (again 😭, y'all are too productive!) to fix `go.mod`/`go.sum` merge conflicts. Have also updated the version numbers in `go.mod`, so that should fix the CI failures.

Is there a better way to handle this? I feel I've been doing a lot of manual `go.mod` maintenance (e.g. merging `require` blocks, fixing version numbers), and running `go mod tidy` after everything, so I feel like I might be missing something!

Have also switched the unit name to be a resource attribute, which I think is a bit more consistent with what the other receivers do.

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev committed:
Fix incorrect version number

SquidDev committed:
Oh, and tidy as well

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

josepcorrea commented:
Hi @SquidDev and reviewers,

This is a fantastic and highly anticipated feature for the `systemdreceiver`. Great work getting this to the testing phase!

I would like to help you and the team by running **functional and integration tests** on my end, specifically focusing on validating the metric output and the behavior of the `resource_to_telemetry_conversion`.

To ensure I'm testing against the correct environment and configuration (especially regarding D-Bus stubbing/mocking, as you mentioned), could you please provide a brief guide or an example configuration for how to:

1.  **Run the receiver with the current branch/PR implementation** in a testing environment (e.g., within a Docker container or against a specific systemd setup).
2.  **Generate the expected unit states** (e.g., forcing a unit into 'failed' or 'reloading' state) to check the binary metric logic.

I will follow up here with any functional test findings, bug reports, or potential suggestions for improvements as soon as I can replicate the setup.

Thanks for your efforts!

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev commented:
@josepcorrea Thanks for helping with this!

> **Run the receiver with the current branch/PR implementation**

 1. Currently the systemd receiver is not built as part of the collector by default. This can be done by adding `gomod: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/systemdreceiver v0.136.0` to the `receivers:` block of `cmd/otelcontribcol/builder-config.yaml`, and then rebuilding the collector (`make genotelcontribcol && make otelcontribcol`).
 2. For testing itself, I have [put together a Dockerfile + config](https://github.com/user-attachments/files/22717809/systemd-test.zip). This includes two services (`otel-test-active`, `otel-test-failed`), which will be active and failed respectively.
 3. The collector can then be started by running `otelcontribcol /etc/otel/config.yaml` inside the container (via a separate `docker exec`).

> **Generate the expected unit states**

So `active/`inactive`/`failed` should be easy (for inactive, just `systemctl stop otel-test-active`). I think the other states are trickier -- you probably create them with a `notify` service type and a short Python script[^1], but I'm not sure it's worth it; at that point you're pretty much just testing systemd rather than the collector!

[^1]: e.g. `reloading` can be created by sending `RELOADING=1` via [`sd_notify`](https://www.freedesktop.org/software/systemd/man/latest/sd_notify.html#), and likewise with deactivating and `STOPPING=1`.

atoulme reviewed (approved):
LGTM, but please help understand why use active_state instead of state?

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev committed:
Update changelog name to match changes in #43215

SquidDev commented on receiver/systemdreceiver/testdata/expected_metrics/basic-scrape.yaml:
```diff
@@ -0,0 +1,145 @@
+resourceMetrics:
+  - resource:
+      attributes:
+        - key: systemd.unit.name
```
I originally had these as metric attributes. However, when adding cgroup support to this receiver, it felt that having a receiver per unit made a bit more sense — this is what we do for the various container receivers for example.

SquidDev commented:
Ahh, sorry! Thought I'd caught all the upstream changes that had gone in (had caught a change to gofumpt), but clearly missed this one. All fixed!

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev commented on receiver/systemdreceiver/metadata.yaml:
```diff
@@ -8,5 +8,37 @@ status:
+      - refreshing
+
+metrics:
+  systemd.unit.state:
```
I think modelling discrete states within OTEL's model is tricky, and definitely something it would be good to have a standard solution for, but I'm not convinced mapping them to integers is the right way here.

 - There's no way to perform aggregates on these values, as the sum or average of two states doesn't give anything meaningful. This isn't the end of the world, but feels a bit of a divergence from standard gauge behaviour.
 - Consuming metrics in this format ends up being pretty awkward, as you're always going to have to consult the lookup table. For instance, if one wanted to have an alert for failing services, your query would look like `systemd.unit.state{} == 3`, which is definitely less self-explanatory than something like `systemd.unit.state{systemd.unit.state="failed"} = 1`!

It would definitely be good to have a better way of handling this. But probably a wider discussion to be had, and as you say, not something we need to do in this PR :).

SquidDev committed:
Make generate

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

pjanotti committed:
Merge branch 'main' into systemd-receiver

pjanotti commented:
@SquidDev - LGTM, let's ensure CI is clean so I can label it `ready to merge` - btw, I just merged main into your branch.

SquidDev committed:
Merge remote-tracking branch 'origin/main' into systemd-receiver

SquidDev commented:
https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/e425f33a598fc047f190c968a5a4c3bb0991333f has immediately introduced conflicts again 😅. Have fixed.
